ooniprobe/TestClasses/MKNetworkTest.mm

```diff
@@ -3,132 +3,257 @@
+        return nil;
+    }
+    return evinfo;
+}
+
 @implementation MKNetworkTest
 
 -(id) init {
     self = [super init];
     if (self) {
         self.backgroundTask = UIBackgroundTaskInvalid;
-        [self createMeasurementObject];
+        self.measurements = [[NSMutableDictionary alloc] init];
+        self.settings = [Settings new];
+        //self.settings.log_filepath = [TestUtility getFileNamed:[self.result getLogFile:self.name]];
     }
     return self;
 }
 
-- (void)createMeasurementObject{
-    self.measurement = [Measurement new];
-    //If needed for the second measurement object (websites)
+- (Measurement*)createMeasurementObject{
+    Measurement *measurement = [Measurement new];
     if (self.result != NULL)
-        [self.measurement setResult_id:self.result];
+        [measurement setResult_id:self.result];
     if (self.name != NULL)
-        [self.measurement setTest_name:self.name];
-    [self.measurement save];
+        [measurement setTest_name:self.name];
+    if (self.reportId != NULL)
+        [measurement setReport_id:self.reportId];
+    [measurement save];
+    return measurement;
 }
 
-- (void)setResultOfMeasurement:(Result *)result{
-    self.result = result;
-    [self.measurement setResult_id:self.result];
+-(void)runTest{
+    self.backgroundTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:^{
+        [[UIApplication sharedApplication] endBackgroundTask:self.backgroundTask];
+        self.backgroundTask = UIBackgroundTaskInvalid;
+    }];
+    dispatch_async(
+                   dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                       NSLog(@"%@", [self.settings getSerializedSettings]);
+                       mk_unique_task taskp{mk_nettest_start([[self.settings getSerializedSettings] UTF8String])};
+                       while (!mk_task_is_done(taskp.get())) {
+                           // Extract an event from the task queue and unmarshal it.
+                           NSDictionary *evinfo = wait_for_next_event(taskp);
+                           if (evinfo == nil) {
+                               break;
+                           }
+                           NSLog(@"Got event: %@", evinfo);
+                           NSString *key = [evinfo objectForKey:@"key"];
+                           NSDictionary *value = [evinfo objectForKey:@"value"];
+                           if (key == nil || value == nil) {
+                               return;
+                           }
+                           if ([key isEqualToString:@"status.started"]) {
+                               //[self updateProgressBar:0];
+                           }
+                           else if ([key isEqualToString:@"status.measurement_start"]) {
+                               NSNumber *idx = [value objectForKey:@"idx"];
+                               NSString *input = [value objectForKey:@"input"];
```

Comment:
For robustness, it's better to _not_ assume that fields will be in the JSON in the future, even though I _should not_ remove them without telling you. So, I would use the following pattern for each branch:

```ObjectiveC
  if ([key isEqualToString:@"<event-name>") {
    NSSomeType *type = [value objectForKey:@"<field-name>"];
    NSSomeOtherType *otherType = [value objectForKey:@"<other-field-name>"];
    // ...
    if (type && otherType /* && ... */) {
      // Either use `type` and `otherType` and `...` directly or pass them to a method
    }
  }
```

In this specific case, hence:

```ObjectiveC
                           else if ([key isEqualToString:@"status.measurement_start"]) {
                             NSNumber *idx = [value objectForKey:@"idx"];
                             NSString *input = [value objectForKey:@"input"];
                             if (idx && input) {
                               Measurement *measurement = [self createMeasurementObject];
                               if ([input length] > 0){
                                 measurement.url_id = [Url getUrl:input];
                                 [measurement save];
                               }
                               [self.measurements setObject:measurement forKey:idx];
                             } else {
                               NSLog(@"Either idx or input is nil");
                               assert(false); // crash if it's a debug build 
                             }
                           }
```

Does this makes sense to you?

Reply:
added null checks

Reply:
Thanks!